test/cpp/microbenchmarks/bm_threadpool.cc

```diff
@@ -0,0 +1,391 @@
+  ThreadPoolAddAnotherHelper(state, 128);
+}
+BENCHMARK(BM_ThreadPool128AddAnother)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool512AddAnother(benchmark::State& state) {
+  ThreadPoolAddAnotherHelper(state, 512);
+}
+BENCHMARK(BM_ThreadPool512AddAnother)->RangePair(524288, 524288, 1, 1024);
+
+static void BM_ThreadPool2048AddAnother(benchmark::State& state) {
+  ThreadPoolAddAnotherHelper(state, 2048);
+}
+BENCHMARK(BM_ThreadPool2048AddAnother)->RangePair(524288, 524288, 1, 1024);
+
+// A functor class that will delete self on end of running.
+class SuicideFunctorForAdd : public grpc_experimental_completion_queue_functor {
+ public:
+  SuicideFunctorForAdd(BlockingCounter* counter) : counter_(counter) {
+    functor_run = &SuicideFunctorForAdd::Run;
+    internal_next = this;
+    internal_success = 0;
+  }
+  ~SuicideFunctorForAdd() {}
+  static void Run(grpc_experimental_completion_queue_functor* cb, int ok) {
+    // On running, the first argument would be itself.
+    auto* callback = static_cast<SuicideFunctorForAdd*>(cb);
+    callback->counter_->DecrementCount();
+    delete callback;
+  }
+
+ private:
+  BlockingCounter* counter_;
+};
+
+// Performs the scenario of external thread(s) adding closures into pool.
+static void BM_ThreadPoolExternalAdd(benchmark::State& state) {
+  static grpc_core::ThreadPool* external_add_pool = nullptr;
+  // Setup for each run of test.
+  if (state.thread_index == 0) {
+    const int num_threads = state.range(1);
+    external_add_pool = grpc_core::New<grpc_core::ThreadPool>(num_threads);
+  }
+  const int num_iterations = state.range(0) / state.threads;
+  while (state.KeepRunningBatch(num_iterations)) {
+    BlockingCounter counter(num_iterations);
+    for (int i = 0; i < num_iterations; ++i) {
+      external_add_pool->Add(new SuicideFunctorForAdd(&counter));
+    }
+    counter.Wait();
+  }
+
+  // Teardown at the end of each test run.
+  if (state.thread_index == 0) {
+    state.SetItemsProcessed(state.range(0));
+    grpc_core::Delete(external_add_pool);
+  }
+}
+BENCHMARK(BM_ThreadPoolExternalAdd)
+    // First pair is range for number of iterations (num_iterations).
+    // Second pair is range for thread pool size (num_threads).
+    ->RangePair(524288, 524288, 1, 1024)
+    ->ThreadRange(1, 256);  // Concurrent external thread(s) up to 256
+
+// Functor (closure) that adds itself into pool repeatedly. By adding self, the
+// overhead would be low and can measure the time of add more accurately.
+class AddSelfFunctor : public grpc_experimental_completion_queue_functor {
+ public:
+  AddSelfFunctor(grpc_core::ThreadPool* pool, BlockingCounter* counter,
+                 int num_add)
+      : pool_(pool), counter_(counter), num_add_(num_add) {
+    functor_run = &AddSelfFunctor::Run;
+    internal_next = this;
+    internal_success = 0;
+  }
+  ~AddSelfFunctor() {}
+  // When the functor gets to run in thread pool, it will take internal_next
```

Comment:
In ThreadPoolWorker::Run(), we pass the functor itself as the first argument, right?